_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,626 @@
+  below); or a new `OP_CHECKSIGADD` opcode may be used to increment a
+  counter if a signature matches a specified public key (e.g. for
+  2-of-3, you need two `OP_CHECKSIGADD` calls to succeed).  This change
+  is made to allow for batch verification of multiple signatures, which
+  [bip-schnorr][] estimates can significantly speed up verification compared to
+  checking each signature independently.
+
+    Ideally, many multisig operations can use signature aggregation as
+    described in the multisig subsection above.  However, it may not be
+    obvious that there are many cases where k-of-n multisig can be done
+    using multiple k-of-k scripts each in a separate Taproot branch.
+    For example, consider the following 2-of-3 multisig Tapscript using
+    `OP_CHECKSIGADD` (`OP_CSADD`):
+
+        <A pubkey> OP_CSADD <B pubkey> OP_CSADD <C pubkey> OP_CSADD 2 OP_EQUAL
+
+    This may be more efficient as three different 2-of-2 multisig
+    scripts in a Taproot merkle tree, allowing whichever one is needed
+    to be chosen:
+
+        (1) <A pubkey> OP_CHECKSIGVERIFY <B pubkey> OP_CHECKSIG
+        (2) <A pubkey> OP_CHECKSIGVERIFY <C pubkey> OP_CHECKSIG
+        (3) <B pubkey> OP_CHECKSIGVERIFY <C pubkey> OP_CHECKSIG
+
+    Similarly, even more efficiency could be gained using key and
+    signature aggregation in the script:
+
+        (1) <AggregatePubkey(<A pubkey>, <B pubkey>)> OP_CHECKSIG
+        (2) <AggregatePubkey(<A pubkey>, <C pubkey>)> OP_CHECKSIG
+        (3) <AggregatePubkey(<B pubkey>, <C pubkey>)> OP_CHECKSIG
+
+- **No direct sigops limit:** because verifying signatures is the most
+  CPU expensive operation in Bitcoin Script, an early version of Bitcoin
+  added a limit on the maximum number of signature-checking operations
+  (sigops) a block could contain, and versions of this limit were
+  applied to later P2SH and segwit v0 scripts.  However, one problem
+  with this limit is that it requires miners consider two factors when
+  trying to select what valid transactions to include in a block:
+  fee density (fee/vbyte) and sigop density (sigops/vbyte).  This is
+  much more difficult than just optimizing block composition based on a
+  single parameter, such as fee density (feerate).
+
+    Taproot resolves this down to one parameter by requiring valid
+    transactions using Taproot spends include a certain amount of data
+    for each sigop that succeeds.  The rule is one free sigop and then
+    50 bytes of data for each additional sigop.  Since Schnorr
+    signatures are at least 64 bytes, this should be more than enough to
+    cover all expected uses.  However, if for some reason a transaction
+    does need to include a higher density of sigops, a special field
+    defined by bip-taproot (the *annex*) currently has no
+    consensus-enforced meaning and so can easily be stuffed with null
+    data to increase the transaction size.  The annex is part of the
```

Comment:
Putting anything in the annex is non-standard (so that it's easy to soft-fork in meaningful stuff later), so it's not _that_ easy to stuff data in there.